File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # memo = []
        memo = [-1] * len(nums)
        return self.func(nums, 0, memo)

    def func (self, arr : List[int], idx : int, dp : List[int]) -> int:

        if (idx >= len(arr)) : return 0

        if (idx == len(arr) - 1) :
            dp[idx] = arr[idx]
            return dp[idx]

        if (dp[idx] != -1) : return dp[idx]

        pick = arr[idx] + self.func(arr, idx + 2, dp)
        notPick = self.func(arr, idx + 1, dp)
        dp[idx] = max(pick, notPick)

        return dp[idx]

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int[] memo = new int[nums.length];
        Arrays.fill(memo, -1);
        return func (nums, 0, memo);
    }

    private int func (int[] nums, int idx, int[] dp){
        if (idx >= nums.length) return 0;

        if (idx == nums.length - 1) {
            dp[idx] = nums[idx];
            return dp[idx];
        }

        if (dp[idx] != -1) return dp[idx];

        int pick = nums[idx] + func(nums, idx + 2, dp);
        int noPick = func(nums, idx + 1, dp);

        dp[idx] = Math.max(pick, noPick);
        return dp[idx];
    }
}